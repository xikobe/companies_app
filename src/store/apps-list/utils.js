/*
 * Creates a normalized object with apps data (including search string)
 * and an array of ids.
 */
export const parseAppsData = (appsList) => {
  const parsedApps = appsList.reduce((result, app) => ({
    apps: {
      ...result.apps,
      [app.id]: {
        ...app,
        searchTerms: `
        ${app.name},
        ${app.description},
        ${app.categories},
        ${app.subscriptions.map((sub) => sub.name)}
        `.toLowerCase(),
        totalPrice: app.subscriptions.reduce((acc, sub) => acc + sub.price, 0),
      },
    },
    ids: [...result.ids, app.id],
  }), { apps: {}, ids: [] });

  parsedApps.ids.sort((a, b) => parsedApps.apps[a].totalPrice - parsedApps.apps[b].totalPrice);

  return parsedApps;
};

/*
 * Parses categories into a normalized object of categories containing ids
 * for corresponding apps. And a sorted array of the available categories.
 */
export const parseCategories = (appsList) => appsList.reduce((result, app) => {
  const byCategory = app.categories.reduce((cats, cat) => ({
    ...cats,
    [cat]: [...(result.appsByCategory[cat] ? result.appsByCategory[cat] : []), app.id],
  }), {});

  return {
    ...result,
    list: [
      ...result.list,
      ...app.categories.filter((category) => !result.list.includes(category)),
    ].sort(),
    appsByCategory: {
      ...result.appsByCategory,
      ...byCategory,
    },
  };
}, { list: [], appsByCategory: {} });
