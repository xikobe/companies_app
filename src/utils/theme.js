import { css } from '@emotion/core';

export const theme = {
  colors: {
    white: '#FFFFFF',
    black: '#172241',
    gray: '#f4f7f6',
    grayDark: '#6c7680',
    teal: '#4bc3c9',
  }
}

export const globalStyles = css`
  @import url('https://fonts.googleapis.com/css?family=Roboto:300,400,500,700,700i');

  body, html {
    font-family: 'Roboto', sans-serif;
    font-weight: 300;
    line-height: 1.5;
    color: ${theme.colors.black};
    text-align: left;
    background-color: ${theme.colors.gray};
  }

  * {
    margin: 0;
    padding: 0;
  }

  /**
  * Allow adding a border to an element by just adding a border-width.
  */
  *,
  *::before,
  *::after {
    border-width: 0;
    border-style: solid;
  }

  ul {
    list-style-type: none;
  }

  p{
    color: ${theme.colors.grayDark};
  }

  h1 {
    color: ${theme.colors.teal};
    font-weight: 400;
  }

  h2 {
    font-size: 2.125rem;
    padding: 0 1rem;
  }

  h3 {
    font-size: 1.5rem;
    padding: 0 .5rem;
    font-weight: 300;
  }
`;
