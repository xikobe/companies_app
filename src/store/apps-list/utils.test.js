import { parseCategories, parseAppsData } from './utils';

const mockApps = [
  {
    id: '9b565b11-7311-5b5e-a699-97873dffb361',
    name: 'Voice Report',
    description: 'Calls reporting and analytics of your calls.',
    categories: ['Dialer', 'Reporting', 'Optimization'],
    subscriptions: [
      {
        name: 'Trial',
        price: 0,
      },
      {
        name: 'Professional',
        price: 3500,
      },
    ],
  },
  {
    id: '470fedc5-489e-5acb-a200-c85adaa18356',
    name: 'Power Dialer',
    description:
      'Auto dialer that will help increase your connect rates and talk time.',
    categories: ['Dialer'],
    subscriptions: [
      {
        name: 'Trial',
        price: 0,
      },
      {
        name: 'Professional',
        price: 4500,
      },
      {
        name: 'Premium',
        price: 6000,
      },
    ],
  },
];

describe('parseAppsData', () => {
  it('returns the correct apps object', () => {
    const parsedApps = parseAppsData(mockApps);

    expect(parsedApps.ids).toEqual(mockApps.map((app) => app.id));
    expect(parsedApps.apps).toMatchSnapshot(); // this is usefull to check big objects
  });
});

describe('parseCategories', () => {
  it('returns the correct categories object', () => {
    const parsedCategories = parseCategories(mockApps);

    expect(parsedCategories.list).toEqual(['Dialer', 'Optimization', 'Reporting']);
    expect(parsedCategories.appsByCategory).toMatchSnapshot();
  });
});
