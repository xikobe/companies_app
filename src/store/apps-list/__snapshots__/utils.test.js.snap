// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAppsData returns the correct apps object 1`] = `
Object {
  "470fedc5-489e-5acb-a200-c85adaa18356": Object {
    "categories": Array [
      "Dialer",
    ],
    "description": "Auto dialer that will help increase your connect rates and talk time.",
    "id": "470fedc5-489e-5acb-a200-c85adaa18356",
    "name": "Power Dialer",
    "searchTerms": "
        power dialer,
        auto dialer that will help increase your connect rates and talk time.,
        dialer,
        trial,professional,premium
        ",
    "subscriptions": Array [
      Object {
        "name": "Trial",
        "price": 0,
      },
      Object {
        "name": "Professional",
        "price": 4500,
      },
      Object {
        "name": "Premium",
        "price": 6000,
      },
    ],
    "totalPrice": 10500,
  },
  "9b565b11-7311-5b5e-a699-97873dffb361": Object {
    "categories": Array [
      "Dialer",
      "Reporting",
      "Optimization",
    ],
    "description": "Calls reporting and analytics of your calls.",
    "id": "9b565b11-7311-5b5e-a699-97873dffb361",
    "name": "Voice Report",
    "searchTerms": "
        voice report,
        calls reporting and analytics of your calls.,
        dialer,reporting,optimization,
        trial,professional
        ",
    "subscriptions": Array [
      Object {
        "name": "Trial",
        "price": 0,
      },
      Object {
        "name": "Professional",
        "price": 3500,
      },
    ],
    "totalPrice": 3500,
  },
}
`;

exports[`parseCategories returns the correct categories object 1`] = `
Object {
  "Dialer": Array [
    "9b565b11-7311-5b5e-a699-97873dffb361",
    "470fedc5-489e-5acb-a200-c85adaa18356",
  ],
  "Optimization": Array [
    "9b565b11-7311-5b5e-a699-97873dffb361",
  ],
  "Reporting": Array [
    "9b565b11-7311-5b5e-a699-97873dffb361",
  ],
}
`;
